### PR TITLE
Publish container images on Docker Hub

### LINKED ISSUE #11986: Publish container images on Docker Hub
### Description

The images provided on https://hub.docker.com/r/jboss/keycloak/tags are out of date, and it is not unusual to [receive requests](https://github.com/keycloak/keycloak/discussions/11235) from the community about publishing those images on Docker Hub.

Considering that, it should be simple to automate with GHA. I believe it is doable to have it.

### Discussion

_No response_

### Motivation

Provide more alternatives to our community

### Details

Considering that we don't have any control over https://hub.docker.com/r/jboss/keycloak, images can be published at https://hub.docker.com/r/keycloak/keycloak/tags

#### Follows up on
- https://github.com/keycloak/keycloak-containers/issues/399

### PR DESCRIPTION
**Pull request by abstractj**
Resolves #11986

### CODE REVIEW

**Comment by abstractj** [comment]
cc @keycloak/kc-maintainers @keycloak/kc-developers

**Comment by andreaTP** [comment]
I will look better into on Mon, but, possibly this can be done using a matrix instead of duplicating steps, or am I missing something?

File: .github/workflows/release-operator-container.yml [hunk 1]
```diff
@@ -11,46 +11,78 @@ env:
...
 
-      - name: Build the operator Docker image
+      - name: Build and push the operator image to Docker
         working-directory: operator
         run: |
-          echo "${{ steps.meta.outputs.tags }}" | xargs -I {} \
+          echo "${{ steps.meta_docker.outputs.tags }}" | xargs -I {} \
           mvn clean package \
               -DskipTests \
+              -Dquarkus.container-image.username=${{ secrets.DOCKER_USERNAME }} \
```
Comment (andreaTP) [inline .github/workflows/release-operator-container.yml:84]: those credentials, "should" not be needed after a successful login.

**Comment by andreaTP** [comment]
As far as I understand this PR I believe that it can be simplified using a `matrix`, please @abstractj feel free to ping me if you need any kind of support in doing so 🙂

**Review by vmuzikar — COMMENTED** [review]
@abstractj Thanks for the fix! +1 for using GHA for pushing to Docker Hub. But AFAIK, we used to use [Docker Hub autobuilds](https://docs.docker.com/docker-hub/builds/). It broke for some reason (I don't have access there to debug). Is it intended to move out of the autobuilds?

**Review by pedroigor — DISMISSED** [review]

File: .github/workflows/release-operator-container.yml (hunk 1 above)
Comment (abstractj) [reply .github/workflows/release-operator-container.yml:84]: @andreaTP I had the chance to experiment with that, before submitting this PR. For some reason, even after a successful login, I was getting HTTP 401. To be honest, I had no clue why. That was the reason why I explicitly included username/password here.

**Comment by abstractj** [comment]
> Sorry for the back and forth @abstractj , I was checking the official documentation and this looks like the suggested path: [docs.github.com/en/actions/publishing-packages/publishing-docker-images#publishing-images-to-docker-hub-and-github-packages](https://docs.github.com/en/actions/publishing-packages/publishing-docker-images#publishing-images-to-docker-hub-and-github-packages)
> 
> basically:
> 
> * login to all the registries
> * configure multiple images to be pushed
> * push them in one step with `docker/build-push-action`
> 
> In general, I'm entirely in favor of this PR:
> 
> * we already have the images, we only need to publish them to another provider
> * ppl complained about not having `keycloak` on docker hub in public threads (e.g. HackerNews)

That was a great finding @andreaTP I will experiment with that for the Keycloak server. Do you have anything in mind for the Keycloak Operator images?

**Comment by abstractj** [comment]
> @abstractj Thanks for the fix! +1 for using GHA for pushing to Docker Hub. But AFAIK, we used to use [Docker Hub autobuilds](https://docs.docker.com/docker-hub/builds/). It broke for some reason (I don't have access there to debug). Is it intended to move out of the autobuilds?

@vmuzikar I believe we moved to GH to centralize all the builds and releases in a single place. But if we agree collectively about moving to auto builds, I don't mind to be honest. The advantage that I see with GHA is the fact that you can easily trigger a new build and override the tags any time we identify a new CVE and want to update the images.
Not sure how it works for auto builds to be honest.

**Review by DGuhr — APPROVED** [review]

File: .github/workflows/release-operator-container.yml (hunk 1 above)
Comment (andreaTP) [reply .github/workflows/release-operator-container.yml]: @abstractj I can see that you removed this change, have you managed to get it to work?

**Comment by andreaTP** [comment]
> That was a great finding @andreaTP I will experiment with that for the Keycloak server. Updated both GHA based on your suggestion.

I have been looking up it, and I think that the best approach would be to use the `tag-push-action` after the first image has been uploaded with `mvn` as shown here:
https://github.com/docker/build-push-action/blob/master/docs/advanced/copy-between-registries.md#copy-images-between-registries

In general, I really like the new encoding that is shaping up, great work @abstractj !

**Review by jonkoops — APPROVED** [review]

File: .github/workflows/release-operator-container.yml (hunk 1 above)
Comment (abstractj) [reply .github/workflows/release-operator-container.yml]: @andreaTP yes, all the changes were tested on my dummy organization on GH, DockerHub and Quay.io.

**Comment by abstractj** [comment]
> I have been looking up it, and I think that the best approach would be to use the `tag-push-action` after the first image has been uploaded with `mvn` as shown here: [https://github.com/docker/build-push-action/blob/master/docs/advanced/copy-between-registries.md#copy-images-between-registries](https://github.com/docker/build-push-action/blob/master/docs/advanced/copy-between-registries.md?rgh-link-date=2022-05-17T07%3A50%3A23Z#copy-images-between-registries)
> 
> In general, I really like the new encoding that is shaping up, great work @abstractj !

Thanks for the feedback @andreaTP. I believe there are many improvements that we can do here, using `tag-push-action` is one of them. I'd like us to move forward with those changes (if they are acceptable), so we can iterate over it.

**Comment by andreaTP** [comment]
sure, LGTM! 👍

**Review by andreaTP — APPROVED** [review]

**Pull request merged**